audinterface/core/utils.py

```diff
@@ -331,3 +334,105 @@ def convert_samples_to_seconds(time):
+):
+    r"""Reshape signal by applying sliding window.
+
+    Args:
+        signal: input signal in shape
+            ``(samples,)``
+            or ``(channels, samples)``
+        sampling_rate: sampling rate in Hz
+        win_dur: window duration,
+            if value is as a float or integer
+            it is treated as seconds.
+            To specify a unit provide as string,
+            e.g. ``'2ms'``.
+            To specify in samples provide as string without unit,
+            e.g. ``'2000'``
+        hop_dur: hop duration,
+            if value is as a float or integer
+            it is treated as seconds.
+            To specify a unit provide as string,
+            e.g. ``'2ms'``.
+            To specify in samples provide as string without unit,
+            e.g. ``'2000'``
+
+    Returns:
+        view of signal with shape ``(frames, channels, samples)``
+
+    Raises:
+        ValueError: if ``win_dur`` or ``hop_dur``
+            is smaller than ``1/sampling_rate``
+
+    Examples:
+        >>> signal = np.array(
+        ...     [
+        ...         [0, 1, 2, 3, 4, 5],
+        ...         [0, 10, 20, 30, 40, 50],
+        ...     ],
+        ... )
+        >>> signal
+        array([[ 0,  1,  2,  3,  4,  5],
+               [ 0, 10, 20, 30, 40, 50]])
+        >>> frames = window(
+        ...     signal,
+        ...     sampling_rate=1,
+        ...     win_dur=3,
+        ...     hop_dur=2,
+        ... )
+        >>> for frame in frames:
+        ...     frame
+        array([[ 0,  1,  2],
+               [ 0, 10, 20]])
+        array([[ 2,  3,  4],
+               [20, 30, 40]])
+
+    """
+    signal = np.atleast_2d(signal)
+
+    win_dur = to_timedelta(win_dur, sampling_rate)
+    hop_dur = to_timedelta(hop_dur, sampling_rate)
+    win_length = int(win_dur.total_seconds() * sampling_rate)
+    hop_length = int(hop_dur.total_seconds() * sampling_rate)
+
+    if win_length <= 0:
+        raise ValueError(
+            f'When the sampling rate is '
+            f'{sampling_rate} '
+            f'Hz the window duration must be at least '
+            f'{1.0/sampling_rate}s, '
+            f'but got '
+            f'{win_dur.total_seconds()}s.'
+        )
+
+    if hop_length <= 0:
+        raise ValueError(
+            f'When the sampling rate is '
+            f'{sampling_rate} '
+            f'Hz the hop duration must be at least '
+            f'{1.0/sampling_rate}s, '
+            f'but got '
+            f'{win_dur.total_seconds()}s.'
+        )
+
+    if signal.shape[1] < win_length:  # signal too short
+        return np.array([], dtype=signal.dtype)
+
+    shape = (signal.shape[0], signal.shape[1] - win_length + 1, win_length)
+    strides = (signal.strides[0], signal.strides[1], signal.strides[1])
+    signal = np.lib.stride_tricks.as_strided(
+        signal,
+        strides=strides,
+        shape=shape,
+    )[:, 0::hop_length]
+
+    signal = signal.swapaxes(0, 1)  # make frames first axis
```

Comment:
Otherwise we would return `(channels, frames, samples)`, but I think `(frames, channels, samples)` makes more sense cause now the user can simply do `frames[idx]` to select a specific frames.

Reply:
In `audinterface.Feature` we request that a process function with a sliding window returns `(channels, features, frames)`, so I guess here `(channels, samples, frames)` would make the most sense?

Reply:
But I also see the point of `for frame in frames`. Maybe you could provide an example how you would use `window` to create an example process function creating the mean per frame.

Reply:
Right, there we have `(channels, features, frames)`. So maybe you are right and it makes more sense to switch the last two axis then.

Reply:
Thanks, so from that example `(frames, channels, samples)` makes sense. On the other hand if we switch to `(channels, features, frames)` we could directly use `sliding_window()` as a processing function.

Reply:
The following should also work, but it is also not very obvious:
```python
y = frames.mean(axis=1).mean(axis=0)
```

But I agree with you that it is better to stick with our convention and return it as `(channels, features, frames)`.

Reply:
Actually, if we return `(channels, features, frames)` the following works:

```
def mean(signal, sampling_rate, win_dur, hop_dur):
    frames = audinterface.utils.sliding_window(
        signal,
        sampling_rate,
        win_dur,
        hop_dur,
    )
    return frames.mean(axis=1)
```

Reply:
And with `frames.mean(axis=1, keepdims=True)` it even works for multiple channels, e.g.:

```python
...
interface = audinterface.Feature(
    'mean',
    process_func=mean,
    process_func_args={
        'win_dur': win_dur,
        'hop_dur': hop_dur,
    },
    win_dur=win_dur,
    hop_dur=hop_dur,
    channels=[0, 0],
)
```
```
                                                                 mean-0    mean-1
file            start                  end                                       
wav/03a01Fa.wav 0 days 00:00:00        0 days 00:00:01        -0.000329 -0.000329
                0 days 00:00:00.500000 0 days 00:00:01.500000 -0.000285 -0.000285
wav/03a01Nc.wav 0 days 00:00:00        0 days 00:00:01         0.000039  0.000039
                0 days 00:00:00.500000 0 days 00:00:01.500000 -0.000412 -0.000412
wav/03a01Wa.wav 0 days 00:00:00        0 days 00:00:01        -0.000284 -0.000284
                0 days 00:00:00.500000 0 days 00:00:01.500000 -0.000318 -0.000318
```

Reply:
I now changed to `(channels, features, frames)` (see description).